onnxruntime/core/optimizer/unsqueeze_elimination.cc

```diff
@@ -2,104 +2,101 @@
 // Licensed under the MIT License.
 
 #include "core/optimizer/unsqueeze_elimination.h"
+#include "core/graph/graph_utils.h"
+#include "core/graph/graph.h"
 
 using namespace ONNX_NAMESPACE;
 using namespace ::onnxruntime::common;
 
 namespace onnxruntime {
 
-Status UnsqueezeElimination::ApplyImpl(onnxruntime::Graph& graph, bool& modified, int graph_level) const {
-  std::vector<onnxruntime::NodeIndex> removed_nodes;
-
-  for (auto& node : graph.Nodes()) {
-    // recurse first as there are early exits in the processing here
-    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level));
-
-    if (node.OpType() != "Unsqueeze" || node.GetInputEdgesCount() != 0 || graph.IsNodeOutputsInGraphOutputs(node)) {
-      continue;
-    }
-
-    const onnxruntime::NodeAttributes& attributes = node.GetAttributes();
-    const ONNX_NAMESPACE::AttributeProto* attr = &attributes.find("axes")->second;
-    if (attr == nullptr || attr->type() != AttributeProto_AttributeType_INTS) {
-      continue;
-    }
+Status UnsqueezeElimination::Apply(Graph& graph, Node& node, bool& modified, bool& removed) {
+  // Get "axes" attribute
+  const ONNX_NAMESPACE::AttributeProto* attr = graph_utils::GetNodeAttribute(node, "axes");
+  if (attr == nullptr || attr->type() != AttributeProto_AttributeType_INTS) {
+    return Status::OK();
+  }
 
-    // Get attribute of "axes"
-    std::vector<int64_t> axes;
-    for (int i = 0; i < attr->ints_size(); i++) {
-      axes.push_back(static_cast<int64_t>(attr->ints(i)));
-    }
+  std::vector<int64_t> axes;
+  for (int i = 0; i < attr->ints_size(); i++) {
+    axes.push_back(static_cast<int64_t>(attr->ints(i)));
+  }
 
-    // Generate new dims
-    NodeArg* input_def = node.MutableInputDefs()[0];
-    const ONNX_NAMESPACE::TensorProto* tensor_proto = nullptr;
-    graph.GetInitializedTensor(input_def->Name(), tensor_proto);
-    if (tensor_proto == nullptr) {
-      continue;
-    }
-    std::vector<int64_t> new_dims(axes.size() + tensor_proto->dims().size(), 0);
-    if (new_dims.size() >= std::numeric_limits<int>::max())
-      return Status(ONNXRUNTIME, FAIL, "index out of range");
+  // Generate new dims
+  NodeArg* input_def = node.MutableInputDefs()[0];
+  const ONNX_NAMESPACE::TensorProto* tensor_proto = nullptr;
+  graph.GetInitializedTensor(input_def->Name(), tensor_proto);
+  if (tensor_proto == nullptr) {
+    return Status::OK();
+  }
+  std::vector<int64_t> new_dims(axes.size() + tensor_proto->dims().size(), 0);
```

Comment:
What's the reason to limit to just Unsqueeze of an initializer?

Reply:
Not sure -- @weixingzhang?

Reply:
The input of shape needs to be updated. For non-initializer Unsqueeze, the shape of input may not there.